packages/context/src/binding-filter.ts

```diff
@@ -9,11 +9,36 @@ import {BindingAddress} from './binding-key';
 /**
  * A function that filters bindings. It returns `true` to select a given
  * binding.
+ *
+ * **NOTE**: We keep the generic type `T` for backward compatibility. It
+ * represents the value type for matched bindings.
  */
-export type BindingFilter<ValueType = unknown> = (
-  binding: Readonly<Binding<ValueType>>,
+// tslint:disable-next-line:no-unused
+export type BindingFilter<T = unknown> = (
+  binding: Readonly<Binding<unknown>>,
 ) => boolean;
 
+/**
+ * A type guard that asserts the value type for a binding if the filter function
+ * returns `true`. This type is very much the same as `BindingFilter` except
+ * that it instructs TypeScript compiler that the `binding` parameter is
+ * `Readonly<Binding<T>>` when the return value is `true`. For example:
+ *
+ * ```ts
+ * const interceptorFilter: BindingFilterGuard<Interceptor> = (
+ *   binding,
+ * ): binding is Readonly<Binding<Interceptor>> => binding.tagMap['interceptor'];
```

Comment:
I believe TypeScript is able to infer types in lambda function arguments and it should be possible to simplify this snippet as follows:

```ts
const interceptorFilter: BindingFilterGuard<Interceptor> = binding => !!binding.tagMap['interceptor'];
```

Reply:
TypeScript treats `(binding) => boolean` and `(binding) => binding is Binding<T>` as incompatible types and they cannot be implicitly casted.